source/core/TypeAndDataPathFinder.cpp

```diff
@@ -0,0 +1,241 @@
+{
+    _tNeedle = tHaystack;
+    _pNeedle = pNeedle;
+    _pHayStack = null;
+    _pathEmpty = true;
+    _found = false;
+    _path = path;
+}
+//------------------------------------------------------------
+//! Visitor dispatch helper method that traverses all the types in a TypeManager
+void PathFinderVisitor::Accept(TypeManagerRef tm)
+{
+    TypeRef type = tm->TypeList();
+    while (type) {
+        Accept(type, type->Begin(kPASoftRead));
+        if (_found)
+            break;
+        type = type->Next();
+    }
+
+    SubString rootName;
+    Boolean isTypeRefConstant = false;
+    if (_found) {
+        rootName = type->Name();
+    } else {
+        _found = tm->PointerToTypeConstRefName((TypeRef*)_pNeedle, &rootName);
+        isTypeRefConstant = _found;
+    }
+
+    if (_found) {
+        if (isTypeRefConstant) {
+            _path->InsertCStr(0, ".");
+        }
+        if (!type->IsA("VirtualInstrument")) {
+            _path->InsertSubString(0, &rootName);
+        }
+    }
+}
+//------------------------------------------------------------
+//! Visitor dispatch helper method that defines a new/smaller haystack
+void PathFinderVisitor::Accept(TypeRef tHayStack, DataPointer pHayStack)
+{
+    if (_pNeedle == tHayStack) {
+        // The value may be a TypeRef. If so check and catch it here,
+        // no need for each visitor to do a check.
+        // _isTypeRef = true;
+        _found = true;
+    } else if ((_pNeedle == pHayStack) && (_tNeedle->IsA(tHayStack, true) )) {
+        // Once the needle is found then there is no need to keep
+        // digging if the types match. However, if the needle is the
+        // first field in a cluster more digging may be necessary
+        // to get the correct full symbol path.
+        _found = true;
+    } else {
+        // Dig deeper into the data.
+        void* saveData = _pHayStack;
+        _pHayStack = pHayStack;
+        tHayStack->Accept(this);
+        _pHayStack = saveData;
+    }
+}
+//------------------------------------------------------------
+void PathFinderVisitor::VisitBad(TypeRef type)
+{
+}
+//------------------------------------------------------------
+void PathFinderVisitor::VisitBitBlock(BitBlockType* type)
+{
+}
+//------------------------------------------------------------
+void PathFinderVisitor::VisitBitCluster(BitClusterType* type)
+{
+}
+//------------------------------------------------------------
+void PathFinderVisitor::VisitCluster(ClusterType* type)
+{
+    if (_pHayStack == null)
+        return;
+
+    IntIndex count = type->SubElementCount();
+    for (IntIndex j = 0; j < count; j++) {
+        TypeRef elementType = type->GetSubElement(j);
+        IntIndex offset = elementType->ElementOffset();
+        AQBlock1* pElementData = (AQBlock1*)_pHayStack + offset;
+        Accept(elementType, pElementData);
+        if (_found) {
+            if (!elementType->ElementName().CompareCStr("Locals")) {
```

Comment:
This seems like it would get false positives -- anyone could name a cluster or cluster elements "Locals".
You only want to identify and avoid the top-level Locals cluster under VI, right?
Could you keep track of search depth or something to make this test more targeted?